File: backend/src/mhglauncher/services/wish_statistics.py
"""祈愿统计算法, 参考 Snap.Hutao 的 TypedWishSummary 实现。"""

from __future__ import annotations

import builtins
from typing import TYPE_CHECKING

from mhglauncher.models import WishBannerDetail, WishBannerItem, WishRecord
from mhglauncher.services.item_metadata import enrich_record

if TYPE_CHECKING:
    from mhglauncher.services.image_cache import ImageCacheService


def guarantee_threshold(gacha_type: str) -> int:
    return 80 if gacha_type == "302" else 90


def build_banner_detail(
    uid: str,
    gacha_type: str,
    items: builtins.list[WishRecord],
    image_cache: ImageCacheService | None = None,
    port: int = 0,
) -> WishBannerDetail:
    threshold = guarantee_threshold(gacha_type)
    orange_pity = 0
    purple_pity = 0
    pull_number = 0
    orange_distances: builtins.list[int] = []
    orange_items: builtins.list[WishBannerItem] = []
    purple_items: builtins.list[WishBannerItem] = []
    five_star_count = 0
    four_star_count = 0
    three_star_count = 0
    max_pity = 0
    min_pity = 0

    for item in items:
        pull_number += 1
        orange_pity += 1
        purple_pity += 1
        rank = item.rank
        if rank == 5:
            item_pity = orange_pity
            if orange_pity < min_pity or min_pity == 0:
                min_pity = orange_pity
            if orange_pity > max_pity or max_pity == 0:
                max_pity = orange_pity
            orange_distances.append(orange_pity)
            orange_pity = 0
            purple_pity = 0
            five_star_count += 1
            enriched = enrich_record(item, image_cache, port)
            orange_items.append(
                WishBannerItem(
                    name=enriched.name,
                    item_id=enriched.item_id,
                    item_type=enriched.item_type,
                    rank=rank,
                    icon_url=enriched.icon_url,
                    pull_number=pull_number,
                    pity=item_pity,
                    time=item.time,
                )
            )
        elif rank == 4:
            item_pity = purple_pity
            purple_pity = 0
            four_star_count += 1
            enriched = enrich_record(item, image_cache, port)
            purple_items.append(
                WishBannerItem(
                    name=enriched.name,
                    item_id=enriched.item_id,
                    item_type=enriched.item_type,
                    rank=rank,
                    icon_url=enriched.icon_url,
                    pull_number=pull_number,
                    pity=item_pity,
                    time=item.time,
                )
            )
        elif rank == 3:
            three_star_count += 1

    total = pull_number
    five_star_percent = five_star_count / total if total > 0 else 0.0
    four_star_percent = four_star_count / total if total > 0 else 0.0
    three_star_percent = three_star_count / total if total > 0 else 0.0
    average_pity = sum(orange_distances) / len(orange_distances) if orange_distances else 0.0

    return WishBannerDetail(
        uid=uid,
        gacha_type=gacha_type,
        total=total,
        time_from=items[0].time if items else None,
        time_to=items[-1].time if items else None,
        five_star_count=five_star_count,
        four_star_count=four_star_count,
        three_star_count=three_star_count,
        five_star_percent=round(five_star_percent, 4),
        four_star_percent=round(four_star_percent, 4),
        three_star_percent=round(three_star_percent, 4),
        max_pity=max_pity,
        min_pity=min_pity if min_pity > 0 else 0,
        average_pity=round(average_pity, 2),
        last_pity=orange_pity,
        last_purple_pity=purple_pity,
        guarantee_threshold=threshold,
        five_star_items=list(reversed(orange_items)),
        four_star_items=list(reversed(purple_items)),
    )

File: backend/src/services/wish-statistics.ts
import type { WishRecord } from "../core/models";
import type { GachaResourceService } from "./gacha-resources";
import { isStandardFiveStar } from "./wish-standard-items";

interface BannerItem {
  name: string; item_id: string; item_type: string; rank: number; icon_url?: string | null;
  pull_number: number; pity: number; time: string;
}
export interface BannerDetail {
  uid: string; gacha_type: string; total: number; time_from: string | null; time_to: string | null;
  five_star_count: number; four_star_count: number; three_star_count: number;
  five_star_percent: number; four_star_percent: number; three_star_percent: number;
  max_pity: number; min_pity: number; average_pity: number; last_pity: number;
  last_purple_pity: number; guarantee_threshold: number; five_star_items: BannerItem[]; four_star_items: BannerItem[];
  // 限定池专属：每个限定五星所需的原石数量依据平均 UP 出金抽数推算；常驻/新手池为 0。
  average_up_pity: number;
  // 限定池专属：小保底不歪率（排除大保底后的 50/50 胜率）；常驻/新手池为 0。
  small_guarantee_win_rate: number;
}

const rounded = (value: number, digits: number): number => Number(value.toFixed(digits));
const LIMITED_TYPES = new Set(["301", "302"]);

export function banner(uid: string, type: string, records: WishRecord[], resources: Pick<GachaResourceService, "enrich">): BannerDetail {
  let orange = 0, purple = 0, three = 0, four = 0, five = 0, max = 0, min = 0;
  let lastUpOrange = 0, smallWin = 0, smallLose = 0;
  let prevIsUp = true; // 第一颗五星之前视为已中，避免首颗即限定被误判为大保底
  const distances: number[] = [], upCycles: number[] = [], oranges: BannerItem[] = [], purples: BannerItem[] = [];
  const limited = LIMITED_TYPES.has(type);
  records.forEach((raw, index) => {
    orange += 1; purple += 1; lastUpOrange += 1;
    const item = resources.enrich(raw);
    if (item.rank === 5) {
      five += 1; distances.push(orange); max = Math.max(max, orange); min = min ? Math.min(min, orange) : orange;
      const isUp = limited && !isStandardFiveStar(item.item_id);
      if (limited) {
        if (isUp) {
          upCycles.push(lastUpOrange);
          // 上一颗非 UP（歪）时本次为触发大保底，不计入小保底统计；
          // 上一颗已为 UP 时本次为正常小保底 50/50 命中。
          if (prevIsUp) smallWin += 1;
        } else if (prevIsUp) {
          // 上一颗已为 UP 时本次为小保底 50/50 失败（歪常驻）。
          smallLose += 1;
        }
        prevIsUp = isUp;
      }
      oranges.push(output(item, index + 1, orange)); orange = 0; purple = 0; lastUpOrange = 0;
    } else if (item.rank === 4) {
      four += 1; purples.push(output(item, index + 1, purple)); purple = 0;
    } else if (item.rank === 3) three += 1;
  });
  const total = records.length;
  const smallTry = smallWin + smallLose;
  return {
    uid, gacha_type: type, total, time_from: records[0]?.time ?? null, time_to: records.at(-1)?.time ?? null,
    five_star_count: five, four_star_count: four, three_star_count: three,
    five_star_percent: rounded(five / (total || 1), 4), four_star_percent: rounded(four / (total || 1), 4),
    three_star_percent: rounded(three / (total || 1), 4), max_pity: max, min_pity: min,
    average_pity: rounded(distances.reduce((a, b) => a + b, 0) / (distances.length || 1), 2),
    last_pity: orange, last_purple_pity: purple, guarantee_threshold: type === "302" ? 80 : 90,
    five_star_items: oranges.reverse(), four_star_items: purples.reverse(),
    average_up_pity: rounded(upCycles.reduce((a, b) => a + b, 0) / (upCycles.length || 1), 2),
    small_guarantee_win_rate: rounded(smallTry ? smallWin / smallTry : 0, 4),
  };
}

function output(item: WishRecord, pull_number: number, pity: number): BannerItem {
  return { name: item.name, item_id: item.item_id, item_type: item.item_type, rank: item.rank, icon_url: item.icon_url, pull_number, pity, time: item.time };
}
